rllib_contrib/alpha_zero/src/rllib_alpha_zero/alpha_zero/mcts.py

```diff
@@ -0,0 +1,157 @@
+
+    @total_value.setter
+    def total_value(self, value):
+        self.parent.child_total_value[self.action] = value
+
+    def child_Q(self):
+        # TODO (weak todo) add "softmax" version of the Q-value
+        return self.child_total_value / (1 + self.child_number_visits)
+
+    def child_U(self):
+        return (
+            math.sqrt(self.number_visits)
+            * self.child_priors
+            / (1 + self.child_number_visits)
+        )
+
+    def best_action(self):
+        """
+        :return: action
+        """
+        child_score = self.child_Q() + self.mcts.c_puct * self.child_U()
+        masked_child_score = child_score
+        masked_child_score[~self.valid_actions] = -np.inf
+        return np.argmax(masked_child_score)
+
+    def select(self):
+        current_node = self
+        while current_node.is_expanded:
+            best_action = current_node.best_action()
+            current_node = current_node.get_child(best_action)
+        return current_node
+
+    def expand(self, child_priors):
+        self.is_expanded = True
+        self.child_priors = child_priors
+
+    def get_child(self, action):
+        if action not in self.children:
+            self.env.set_state(self.state)
+            obs, reward, terminated, truncated, _ = self.env.step(action)
+            next_state = self.env.get_state()
+            self.children[action] = Node(
+                state=next_state,
+                action=action,
+                parent=self,
+                reward=reward,
+                done=terminated,
+                obs=obs,
+                mcts=self.mcts,
+            )
+        return self.children[action]
+
+    def backup(self, value):
+        current = self
+        while current.parent is not None:
+            current.number_visits += 1
+            current.total_value += value
+            current = current.parent
+
+
+class RootParentNode:
+    def __init__(self, env):
+        self.parent = None
+        self.child_total_value = collections.defaultdict(float)
+        self.child_number_visits = collections.defaultdict(float)
+        self.env = env
+
+
+class MCTS:
+    def __init__(self, model, mcts_param):
+        self.model = model
+        self.temperature = mcts_param["temperature"]
+        self.dir_epsilon = mcts_param["dirichlet_epsilon"]
+        self.dir_noise = mcts_param["dirichlet_noise"]
+        self.num_sims = mcts_param["num_simulations"]
+        self.exploit = mcts_param["argmax_tree_policy"]
+        self.add_dirichlet_noise = mcts_param["add_dirichlet_noise"]
+        self.c_puct = mcts_param["puct_coefficient"]
+
+    def compute_action(self, node):
+        for _ in range(self.num_sims):
+            leaf = node.select()
+            if leaf.done:
+                value = leaf.reward
+            else:
+                child_priors, value = self.model.compute_priors_and_value(leaf.obs)
+                if self.add_dirichlet_noise:
```

Comment:
Are we sure we want to add dirichlet noise to every child prior, not only to the ones starting in the root node? This behaviour differs from the one proposed in the paper and the [open spiel implementation](https://github.com/google-deepmind/open_spiel/blob/77a03df63f6e6ddbf15ccf22e7f1ca5ff841dd2e/open_spiel/python/algorithms/mcts.py#L316)